### ISSUE TITLE
Functions curve editor hard to use due to small channels list  and other regressions from harlequin

### ISSUE THREAD

**Issue by blurymind** [opentoonz/opentoonz#234]
here is the function editor in toonz harlequin:
![functionseditor-regressions](https://cloud.githubusercontent.com/assets/6495061/14783726/84856848-0ae9-11e6-8b9d-0687ff09c331.JPG)

and here it is in opentoonz:
![layout-graph](https://cloud.githubusercontent.com/assets/6495061/14783529/303b8a0c-0ae8-11e6-9d55-282276f2ff7f.JPG)
- The biggest problem is the much lesser vertical space for the channels list. This actually undermines workflow for both curves and keyframe spreadsheet editing.
- you need to have the function editor constantly visible in order to use the curves editor. I think that is a problem too. It's good that we can now split them - but it's not good that we need to have them both take up precious screen estate when most of the time in many cases people need only one or the other open.
- The context sensitive editing area that shows values has been partially removed in opentoonz.
  It no longer works on curves or with the curves editor at all. 
  That was a very useful feature for viewing and editing selected curves and keyframe values. 
  It's one that if developed further could allow us to execute math operations on a curve - the way both maya and blender's graph editors can do (*-1 on y would flip a curve vertically for example, *1.5 horizontally would make it 50% slower ) 

Please note how small the channels list is in opentoonz as compared to toonz harlequin. 
for people with one monitor and especially people on laptops this makes it quite unwieldy to use.
When working with curves, we need this list to be visible and to show as many as possible.
Very often the animator has to find curves and toggle their visibility on and off. This is done through that list.

We do this so often that some other software such as blender even have a 'filter' search box to make the process easier:

https://wiki.blender.org/index.php/User:Tnboma/Doc:2.5/Manual/Animation/Graph_Editor#Display_Filtering

> You can filter F-Curves by channel names or text by clicking the Magnifying Glass icon, and entering text. Only channels containing this text will be displayed.

**Comment by blurymind** [comment]
the tween control widgets from the x-sheet in the curves editor is another lost feature:

toonz harlequin
![functionseditor-spacingwidget](https://cloud.githubusercontent.com/assets/6495061/14749542/dbb5036e-08b7-11e6-9c49-cdf7968e71f7.gif)

opentoonz
![missingcurvesfeature](https://cloud.githubusercontent.com/assets/6495061/14836054/5c108bd2-0c05-11e6-979c-b899048a3301.JPG)

This allowed to move all tween widgets that are aligned in time together - something that can not be done in the x-sheet.
page 365
http://www.toonz.com/cgi-shl/download/71H/Toonz%20Harlequin%2071%20User%20Guide.pdf

**Comment by psychoanima** [comment]
Well, tween controls are definitely not lost in my version of OpenToonz.

![capture curves](https://cloud.githubusercontent.com/assets/18187104/14951474/763a5bca-1059-11e6-84e5-7e045aef0df0.PNG)


You have to change interpolation in order to see them...

**Comment by blurymind** [comment]
oh sorry, maybe i got that wrong. I will give it a go

**Comment by blurymind** [comment]
@psychoanima I cant get the handles to show anywhere in 1.0.2 release. They are completely gone now.

Please see:
https://github.com/opentoonz/opentoonz/issues/270
Seems like a regression from last nightly build.

**Comment by blurymind** [comment]
@psychoanima can you give me a step by step way to make the handles appear both in x-sheet and curves editor? Are you using 1.0.2?

**Comment by Juls3000** [comment]
@blurymind here are my observations:

(1) Handles work whith Ease in/out interpolation.
(2) Interactivity works in this way for keyframes and handles
(3) Interactivity works in this way just for keyframes (editing handles you lost them in Xsheet)
(4) Sometimes you need Ctlr+drag to move this.
(5) You can edit here the values (keyframes ot handles) of several elements at once when all these values are aligned and have a common control here.

![cats](https://cloud.githubusercontent.com/assets/19145844/14955894/42f31ca6-107d-11e6-838b-5506b05a37d9.jpg)

**Comment by psychoanima** [comment]
@blurymind 

Sure, here's how to do it:


1. Make animation with at least two keyframes

2. Open **Function editor** & **Function curves**

3. To be able to see keyframes in the Function editor and Function curves, you have to select Column from the **Stage** tree. In my case I select **Col1** folder and my keyframes are shown as on picture bellow:

![001](https://cloud.githubusercontent.com/assets/18187104/14955808/c71c005c-107c-11e6-8b67-485d3fbc54c0.JPG)

Now, I can't see small controls in Function curves that can help me to modify interpolation between keyframes without moving curve handles


To be able to see them, you have to have selected only one channel from Stage tree. In my case I will deselect all channels except N/S. And voila! Controls become available.

![002](https://cloud.githubusercontent.com/assets/18187104/14955916/5d338614-107d-11e6-8233-49a3ccdbcefe.JPG)


In case you still don't see them, change interpolation for the keyframes in Function editor. Select the column where the keyframe is and from the drop down menu select appropriate interpolation (ease in/out, speed in/out...) and click Apply

![003](https://cloud.githubusercontent.com/assets/18187104/14955950/9975c9ac-107d-11e6-9300-a08977b7195c.jpg)

**Comment by blurymind** [comment]
Ah I see - my animation in the preferences was set to speed in / speed out for some reason. Thats why it didnt show

**Comment by blurymind** [comment]
There is a problem however - if I move the handle it dissappears from the x-sheet for some reason:
![handledisap](https://cloud.githubusercontent.com/assets/6495061/14957090/4479f024-107b-11e6-96ba-b79d1757d61e.gif)

compare it to this:
![functionseditor-spacingwidget](https://cloud.githubusercontent.com/assets/6495061/14749542/dbb5036e-08b7-11e6-9c49-cdf7968e71f7.gif)

**Comment by Juls3000** [comment]
Yes, it seems that in TH hadles are syncronized in two ways (Funtion editor > Xsheet, and viceversa) but in OT, that happen just in one way (Xsheet > Function editor). It is that I was trying to explain in my former message above.

**Comment by blurymind** [comment]
So basically that is the regression from harlequin

**Comment by psychoanima** [comment]
@blurymind did you read my post above? 

> Ah I see - my animation in the preferences was set to speed in / speed out for some reason. Thats why it didnt show

This has nothing to do with the problem. Both interpolation have same controls that you found "missing". Pls read my step by step instruction above.

> There is a problem however - if I move the handle it dissappears from the x-sheet for some reason

It is not problem in the software. Handles are facing increments differently in x sheet and in function curves. In xsheet they depends only on timing (frames), in Function curves they depends on timing and spacing respectively (you can't move handle in Function curves to 0.7 and expect to be visible in xsheet where frames are incremented by one).

**Comment by Juls3000** [comment]
I suppose, but must confess I am not familiarized too much with editing functions courves. :(

**Comment by Juls3000** [comment]
`It is not problem in the software. Handles are facing increments differently in x sheet and in function curves. In xsheet they depends only on timing (frames), in Function curves they depends on timing and spacing respectively (you can't move handle in Function curves to 0.7 and expect to be visible in xsheet where frames are incremented by one).`

Good point !

**Comment by Juls3000** [comment]
So that I see it in that way: meanwhile it is enough with handles in xSheet no manage animation profile do it, if you have to go to function editor, then forget Sheet to manage handles. Am I right ?

**Comment by psychoanima** [comment]
@Juls3000
 No. Xsheet is showing all keyframes for selected object. 
Function editor, however, can isolate (switch off) keyframed channels that you don't want to see (like in my step by step example above) 

What's happening with your example guys is that keyframes (and handles) in xsheet are overlapping and that's why they are disappearing. So, if X and Y transformation handles for example are overlapping in Xsheet, OpenToonz  will hide them to avoid animators errors (like moving handle from unwanted keyframe which is sharing the same frame position with aimed one).
This is where function curves comes handy, to isolate which keyframe from which channel you want to modify.

**remember: everytime keyframe handles are overlapping in xsheet, that handles are gonna disappear. The reason is practical: to avoid unwanted transformations**

**Comment by Juls3000** [comment]
@psychoanima that have all the sense :) by Millions  and  billions  of  blue blistering  barnacles !!!. So that, you think that in v1.0.2, Funstion editor works ok ?

**Comment by psychoanima** [comment]
@Juls3000 Yup! :) In my opinion everything works great in Function editor. Just keep in mind that when two handles are overlapping (either in xsheet or function curves) they gonna disappear. If that happens open function curves, isolate the channel and "set it free" :)

**Comment by Juls3000** [comment]
great! Thank you again!

**Comment by blurymind** [comment]
The interface is still pretty bad if you are using a laptop though :) 

Thank you for the explanation.
It's strange that when using harlequin I got these handles to show right away - without all the extra steps explained here

**Comment by blurymind** [comment]
I am pretty sure it's a bug or a missing feature :)
Just compare harlequin and OT, do these steps on both:

1. make a level, extend it to 20 frames in the x-sheet
2. set a key on frame 1 go to frame 20 and move the level with the edit tool (E) - this will set another translation key
3. At this point you will see the handles in x-sheet and in the curves editor
4. ctrl+c and ctrl+v the level in x-sheet - this will make a copy of the level, like this:
 
![handle](https://cloud.githubusercontent.com/assets/6495061/15260320/1f5b9d48-194e-11e6-80b3-56b474d465ed.JPG)

make sure the both levels are exposed to the curves editor (circled them in red)
the handle in the curves editor is on 14, move it to 15 or 16, as you move it, The handles on both levels in the x-sheet will move as well!

Now lets repeat the process in Open toonz!
At step 3 - no handles appear at neither locations

so I set it to speed in/out instead of ease in/out. Delete the keyframes and repeat the steps.

Interpolation speed in/out, still no handles anywhere:
![nohandles](https://cloud.githubusercontent.com/assets/6495061/15260688/24928b30-1950-11e6-8ed2-7143ad7c2cf4.JPG)

So why is this so hit and miss in open toonz and just works out of the box in harlequin?
There us definitely a bug in there.

**Comment by blurymind** [comment]
Ok, here we have not one but a couple of bugs, connected with handles.
Since this thread has become long and hard to follow, I am going to split the one that I know for sure why happens.
@psychoanima  helped me identify it.
https://github.com/opentoonz/opentoonz/issues/300

This is not supposed to happen by design and it's not how it works in toonz harlequin.

**Comment by blurymind** [comment]
@BrundleThwaite  in your bug post we need to identify why the handles dissapear in the x-sheet!

**Comment by blurymind** [comment]
I found out that it's the edit tool!
![globalkey](https://cloud.githubusercontent.com/assets/6495061/15261815/6cb53920-1956-11e6-97b8-739e9859284f.JPG)

Ghibli had removed 'Global key' option entirely. This was by default on in Harlequin.
When you start animating, setting a keyframe on frame one (key icon) will create a kf on all the channels.
Then after that when you pose the level, another keyframe will only be created on translation rotation or whatever you do.
This causes the handles to disappear while all the channels are visible. They disappear in both the x-sheet and the curves editor. That is why global key was available before and it was on by default.

as a result, when you repeat the steps, you get missing handles in

**Comment by blurymind** [comment]
Ok the handles issue is resolved now, but we detracted from the main issue here - the interface changes. The interface layout for the curves editor is much worse than the one in toonz harlequin as stated in the first post

**Comment by turtletooth** [comment]
@shun-iwasawa Is the spreadsheet view in the function editor new to the Ghibli version, or was it in Toonz also?  Why was the spreadsheet made the default view instead of the graph?  How about an option that allows the user to choose between spreadsheet or graph as the default view?

**Comment by shun-iwasawa** [comment]
@turtletooth In Toonz Harlequin, the spreadsheet and the graph were alternative. They were stacked in the panel using `QStackedWidget`. I bought out the graph view and put it into the popup, since the staffs in Ghibli demanded to observe the both views at the same time. 
I left the spreadsheet in the main panel because, again, Ghibli staffs preferred to do so. IMO they would like to put the spreadsheet in the main panel for the following reasons:

1. Interface compatibility with Xsheet: Function Editor is usually used in composite work along with Xsheet and Schematic. The spreadsheet view is designed to be put side by side with Xsheet panel, i.e. the direction of time axis, raw height and scroll position are synchronized. Putting the spreadsheet into popup will lose such compatibility. 

1. Precise control: It was difficult to edit the values precisely in the graph view.

1. Consistency with older version: Toonz 4.6 had no graph view. So it's more familiar for them to work on spreadsheet.

AFAIK the graph view was used for a limited purpose in Ghibli - for adjusting curve shape with handles and for moving the keyframe to negative.

Why you would like to make the graph view the default? Will it be easy for beginners..?
Anyway, it is OK to make the option to let the graph to be the default view. Maybe the easy implementation would be just swap the two, putting the spreadsheet into popup and putting the graph into panel.

**Comment by turtletooth** [comment]
@shun-iwasawa Thanks for the detailed answer.

I personally prefer using a graph editor for adjusting curves as it is a more visual representation of what is happening in the animation and I think that it makes the software more familiar to people familiar with the curves in Flash, After Effects and ToonBoom.  

Tonight I played with putting the graph editor in the main function editor window and sometimes it seemed like it made playback go slower.  Have you noticed any performance hit when the graph is opened?  It wasn't all the time and could be just my perception.

I agree with the Ghibli crew that seeing both at the same time is useful along with the other reasons you mentioned.

The benefits to the graph view are the ability to see more at once than the spreadsheet allows and the ease of changing values and curves.

I played with making the spreadsheet values change with a middle click, but there are variety of different value types to consider with that, but I think it would be useful since now changing a value involves double clicking and editing the value.

I can also see @blurymind 's perspective about the space being a little hard to use and getting to the graph editor is a little cumbersome, but I'm not sure what the solution is.

**Comment by turtletooth** [comment]
Based on the discussion in #1207 feature requests are going to be closed here on GitHub unless there is a developer actively working on the feature. Feature requests can be discussed at the OpenToonz Google Group: https://groups.google.com/forum/#!forum/opentoonz_en
